module/zfs/vdev_label.c

```diff
@@ -208,14 +208,77 @@ vdev_label_write(zio_t *zio, vdev_t *vd, int l, void *buf, uint64_t offset,
 }
 
 /*
+ * Generate the nvlist representing this vdev's stats
+ */
+static void
+vdev_config_generate_stats(vdev_t *vd, nvlist_t *nv)
+{
+	vdev_stat_t *vs;
+	vdev_stat_ex_t *vsx;
+
+	vs = kmem_alloc(sizeof (*vs), KM_SLEEP);
+	ASSERT(vs);
+
+	vsx = kmem_alloc(sizeof (*vsx), KM_SLEEP);
+	ASSERT(vsx);
+
+	vdev_get_stats_ex(vd, vs, vsx);
+	fnvlist_add_uint64_array(nv, ZPOOL_CONFIG_VDEV_STATS,
+			(uint64_t *)vs, sizeof (*vs) / sizeof (uint64_t));
+
+	kmem_free(vs, sizeof (*vs));
+
+	/* Add extended stats nvlists */
+	fnvlist_add_uint64_array(nv, ZPOOL_CONFIG_QUEUE_ACTIVE,
+	    vsx->vsx_queue_active, ARRAY_SIZE(vsx->vsx_queue_active));
```

Comment:
I think the per-queue nvlists are a good idea.  I'll update the code for that.

I wasn't aware of the nvlist of nvlists.  Sounds like a nice way to group the stats.  Let me look into how that will work with the existing code before I commit to it.